### PR TITLE
Last major set of ORT format model changes

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Add build setup for minimal build
Group some of the build settings so binary size reduction options are all together
Make some cmake variable naming more consistent
Replace usage of std::hash with murmurhash3 for kernel. std::hash is implementation dependent so can't be used.
Add initial doco and ONNX to ORT model conversion script
Misc cleanups of minimal build breaks.

**Motivation and Context**
ORT format build

### CODE REVIEW

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -698,14 +699,17 @@ if(NOT onnxruntime_USE_FULL_PROTOBUF)
...
+  endif()
+  set_target_properties(onnx PROPERTIES FOLDER "External/ONNX")
+  set_target_properties(onnx_proto PROPERTIES FOLDER "External/ONNX")
+else()
+  include(onnx_minimal)
 endif()
-set_target_properties(onnx PROPERTIES FOLDER "External/ONNX")
-set_target_properties(onnx_proto PROPERTIES FOLDER "External/ONNX")
-
 if (onnxruntime_USE_MKLML)
```
Comment (skottmckay) [inline cmake/CMakeLists.txt:713]: Will refactor this a little as the calls to target_compile_definitions and set_target_properties are duplicated in onnx_minimal.cmake.

**Comment by gwang-msft** [comment]
>     add_link_options(-Wl,--gc-sections)

Here is the setting for apple

    if (CMAKE_HOST_SYSTEM MATCHES "Darwin")
      add_link_options(-Wl -dead_strip)
    else()
      add_link_options(-Wl --gc-sections)
    endif()

---
Refers to: cmake/CMakeLists.txt:224 in fa3e2aa. [](commit_id = fa3e2aa050717c1ca9fa6eed1a57ba6cd8ab9942, deletion_comment = False)

File: cmake/external/onnx_minimal.cmake [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      "${ONNX_SOURCE_ROOT}/onnx/*.cc"
+  )
+  file(GLOB_RECURSE onnx_exclude_src CONFIGURE_DEPENDS
+      "${ONNX_SOURCE_ROOT}/onnx/py_utils.h"
+      "${ONNX_SOURCE_ROOT}/onnx/proto_utils.h"
+      "${ONNX_SOURCE_ROOT}/onnx/backend/test/cpp/*"
+      "${ONNX_SOURCE_ROOT}/onnx/test/*"
+      "${ONNX_SOURCE_ROOT}/onnx/cpp2py_export.cc"
+  )
+  list(REMOVE_ITEM onnx_src ${onnx_exclude_src})
```
Comment (gwang-msft) [inline cmake/external/onnx_minimal.cmake:49]: Remove this and add failure here something like,
if(NOT onnxruntime_MINIMAL_BUILD)
  message(FATAL_ERROR "onnxruntime_MINIMAL_BUILD required to build onnx_minimal")

Comment (skottmckay) [reply cmake/external/onnx_minimal.cmake:49]: we don't include this file unless it's a minimal build though

File: tools/python/convert_onnx_model_to_ort.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    # Second, convert optimized ONNX model to ORT format
+    so.optimized_model_filepath = ort_target_path
+    so.add_session_config_entry('session.save_model_format', 'ORT')
+
+    print("Converting optimized ONNX model {} to ORT format model {}".format(onnx_target_path, ort_target_path))
+    _ = ort.InferenceSession(onnx_target_path, sess_options=so)
+
+    orig_size = os.path.getsize(onnx_target_path)
+    new_size = os.path.getsize(ort_target_path)
+    print("Serialized {} to {}. Sizes: orig={} new={} Diff={} Ratio={:.4f}%".format(
```
Comment (gwang-msft) [inline tools/python/convert_onnx_model_to_ort.py:33]: >Ratio={:.4f}% [](start = 62, length = 13)

This will output 150% as 1.5%, use {:4%}?

Comment (skottmckay) [reply tools/python/convert_onnx_model_to_ort.py:33]: Changed to output 'new:old={:.4f}:1.0'

---
In reply to: [483306631](https://github.com/microsoft/onnxruntime/pull/5056#discussion_r483306631) [](ancestors = 483306631)

File: docs/ONNX_Runtime_Format_Model_Usage.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+##### Linux
+
+`<ORT repository root>/build.sh --config=MinSizeRel --build_shared_lib --minimal_build --disable_rtti --disable_ml_ops --disable_exceptions`
+
+## Executing ORT format models
+
+The API for executing ORT format models is the same as for ONNX models. See the [ORT API documentation](https://github.com/Microsoft/onnxruntime/#api-documentation).
+
+If you provide a filename for the ORT format model, a file extension of '.ort' will be inferred to be an ORT format model.
```
Comment (manashgoswami) [inline docs/ONNX_Runtime_Format_Model_Usage.md:72]: is it _.optimized.onnx_ or _.ort_?

File: docs/Reduced_Operator_Kernel_build.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
+# ONNX Runtime Reduced Operator Kernel build
+
+In order to reduce the compiled binary size of ONNX Runtime (ORT), the operator kernels included in the build can be reduced to just the kernels required for your scenario. 
+
+The kernels to include must first be identified, and secondly the ORT kernel registration source files must be updated to exclude the unused kernels. 
+
+Finally ORT must be manually built.
+
+When building ORT with a reduced set of kernel registrations, `--skip_tests` *MUST* be specified as the kernel reduction will render many of the unit tests invalid.
```
Comment (manashgoswami) [inline docs/Reduced_Operator_Kernel_build.md:9]: Can this be merged with the [ONNX_Runtime_Format_Model_Usage.md](ONNX_Runtime_Format_Model_Usage.md) to include the build with the build steps there?
OR
Move the build steps here and leave that doc for using ORT for Mobile platform.

File: docs/ONNX_Runtime_Format_Model_Usage.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
+# ONNXRuntime Format Model Usage
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply docs/ONNX_Runtime_Format_Model_Usage.md:1]: .optimized.onnx is an ONNX format model that is saved after ORT optimizations run. Different to a .ort model. We want to optimize the ONNX model first as we can't do that at runtime in a minimal build.

I don't want to be too specific about 'flatbuffers' as we could change what we use in the future for an 'ORT format' model.

File: docs/ONNX_Runtime_Format_Model_Usage.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+Perform the following steps to create a minimal build from an ONNX format model.
+
+### Create ORT format model
+
+We will use a helper python script to convert an existing ONNX format model into an optimized ORT format model.
+This will require the ORT python package to be installed, and the ORT repository to have been cloned.
+
+  - Install the ORT nighly python package from https://test.pypi.org/project/ort-nightly/
+  - Convert the ONNX model to ORT format
+    - `python <path to ORT repository>/tools/python/convert_onnx_model_to_ort.py <path to .onnx model>`
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply docs/ONNX_Runtime_Format_Model_Usage.md:22]: This script is one at a time currently.

File: docs/ONNX_Runtime_Format_Model_Usage.md (hunk 4 above)
Comment (skottmckay) [reply docs/ONNX_Runtime_Format_Model_Usage.md:72]: .ort. There are two steps in model conversion. ONNX -> ONNX + ORT optimizations (.optimized.onnx) -> ORT format (.ort)

File: docs/Reduced_Operator_Kernel_build.md (hunk 5 above)
Comment (skottmckay) [reply docs/Reduced_Operator_Kernel_build.md:9]: They should be separate. A reduced ops build is the 'medium' size ORT and is not dependent on the minimal build (which is the 'small' size ORT), so there's only one step to do that. For the minimal build there are multiple steps. 

I could maybe add an example command line to call the reduction script to the minimal build doc so there's less need to fully read this document when doing a minimal build.

File: docs/ONNX_Runtime_Format_Model_Usage.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    - This script will first optimize the ONNX model and save it with a '.optimized.onnx' file extension
+      - this optimized model should be used as input to the reduced build
+    - It will next convert the optimized ONNX model to ORT format.
+
+### Setup information to reduce build to minimum set of operator kernels required
+
+In order to reduce the operator kernels included in the build the required set must be either inferred from one or more ONNX models, or explicitly specified.d
+
+To infer, put one or more ONNX models in a directory. The directory will be recursively searched for '.onnx' files. If taking this approach, you should only include the optimized ONNX models and not both the original and optimized models.
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply docs/ONNX_Runtime_Format_Model_Usage.md:31]: It should be a separate folder. If you run the reduction script against where you did the conversion you will have the original onnx model and the optimized one. That will most likely result in more kernels than you need being included if the optimization changed the graph. e.g. if the model had a MatMul and Add node and we merged that into a Gemm when optimizing you may not need MatMul and Add kernels anymore.

**Comment by gwang-msft** [comment]
>   string(APPEND CMAKE_C_FLAGS " /EHsc /wd26812")

Remove these 2?

---
Refers to: cmake/CMakeLists.txt:327 in fa3e2aa. [](commit_id = fa3e2aa050717c1ca9fa6eed1a57ba6cd8ab9942, deletion_comment = False)

File: cmake/external/onnx_minimal.cmake [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+endif()
+
+if (WIN32)
+    target_compile_options(onnx PRIVATE
+        /wd4800 # 'type' : forcing value to bool 'true' or 'false' (performance warning)
+        /wd4125 # decimal digit terminates octal escape sequence
+        /wd4100 # 'param' : unreferenced formal parameter
+        /wd4244 # 'argument' conversion from 'google::protobuf::int64' to 'int', possible loss of data
+        /EHsc   # exception handling - C++ may throw, extern "C" will not
+        /wd4996 # 'argument' Using double parameter version instead of single parameter version of SetTotalBytesLimit(). The second parameter is ignored.
```
Comment (gwang-msft) [inline cmake/external/onnx_minimal.cmake:76]: Remove this line?

File: docs/ONNX_Runtime_Format_Model_Usage.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (skottmckay) [reply docs/ONNX_Runtime_Format_Model_Usage.md:1]: Attempted to clarify. Renamed file and updated header.

File: docs/ONNX_Runtime_Format_Model_Usage.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+The minimal build can be used with any ORT format model, provided that the kernels for the operators used in the model were included. i.e. the custom build provides a set of kernels, and if that set satisfies a given ORT format model's needs the model can be loaded and executed. 
+
+## Steps to create model and minimal build
+
+You will need a script from the the ORT repository, and to perform a custom build, so you will need to clone the repository locally. See [here](https://github.com/microsoft/onnxruntime/blob/master/BUILD.md#prerequisites) for initial steps.
+
+Perform the following steps to create a minimal build from an ONNX format model.
+
+### Create ORT format model
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply docs/ONNX_Runtime_Format_Model_Usage.md:15]: Attempted to clarify.
Numbered steps.

File: docs/ONNX_Runtime_Format_Model_Usage.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (skottmckay) [reply docs/ONNX_Runtime_Format_Model_Usage.md:22]: Added clarification

File: docs/ONNX_Runtime_Format_Model_Usage.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+We will use a helper python script to convert an existing ONNX format model into an optimized ORT format model.
+This will require the ORT python package to be installed, and the ORT repository to have been cloned.
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply docs/ONNX_Runtime_Format_Model_Usage.md:18]: Added

File: docs/ONNX_Runtime_Format_Model_Usage.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (skottmckay) [reply docs/ONNX_Runtime_Format_Model_Usage.md:31]: Added more details.

File: docs/Reduced_Operator_Kernel_build.md (hunk 5 above)
Comment (skottmckay) [reply docs/Reduced_Operator_Kernel_build.md:9]: Added example to ORT Mobile doco

File: cmake/external/onnx_minimal.cmake (hunk 9 above)
Comment (skottmckay) [reply cmake/external/onnx_minimal.cmake:76]: Removed

File: tools/python/convert_onnx_model_to_ort.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
+import argparse
```
Comment (pranavsharma) [inline tools/python/convert_onnx_model_to_ort.py:1]: License header

File: onnxruntime/core/framework/murmurhash3.cc [hunk 13]
```diff
@@ -0,0 +1,311 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/framework/murmurhash3.h"
+
+// Original source: https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp
```
Comment (pranavsharma) [inline onnxruntime/core/framework/murmurhash3.cc:6]: Also add to third party notices file.

**Review by pranavsharma — DISMISSED** [review]

**Review by pranavsharma — APPROVED** [review]

**Comment by snnn** [comment]
@skottmckay, do you remember why we needed to create a custom cmake file for ONNX?

**Pull request merged**